model/modelprocessor/datastream.go

```diff
@@ -40,40 +40,43 @@ const (
 	rumTracesDataset = "apm.rum"
 )
 
-// SetDataStream is a model.BatchProcessor that routes events to the appropriate
+// SetDataStream is a modelpb.BatchProcessor that routes events to the appropriate
 // data streams.
 type SetDataStream struct {
 	Namespace string
 }
 
 // ProcessBatch sets data stream fields for each event in b.
-func (s *SetDataStream) ProcessBatch(ctx context.Context, b *model.Batch) error {
+func (s *SetDataStream) ProcessBatch(ctx context.Context, b *modelpb.Batch) error {
 	for i := range *b {
+		if (*b)[i].DataStream == nil {
+			(*b)[i].DataStream = &modelpb.DataStream{}
+		}
 		(*b)[i].DataStream.Namespace = s.Namespace
 		if (*b)[i].DataStream.Type == "" || (*b)[i].DataStream.Dataset == "" {
-			s.setDataStream(&(*b)[i])
+			s.setDataStream((*b)[i])
 		}
 	}
 	return nil
 }
 
-func (s *SetDataStream) setDataStream(event *model.APMEvent) {
-	switch event.Processor {
-	case model.SpanProcessor, model.TransactionProcessor:
+func (s *SetDataStream) setDataStream(event *modelpb.APMEvent) {
+	switch {
+	case event.GetProcessor().IsSpan(), event.GetProcessor().IsTransaction():
```

Comment:
The "processor" fields are a bit of a relic, and we should aim to remove them in the long term. With that in mind, I wonder if we should change the model a little bit so we can remove them from the apm-data codebase, and set the `processor.*` fields in our ingest pipelines, or by setting a value on `constant_keyword` fields where it makes sense.

e.g. for metrics, `processor.name` and `processor.event` are both always "metric", so we can update their field definitions to set the value in the mapping: https://github.com/elastic/apm-server/blob/23fb1577909836ebf45e65705df3fd560de5adb1/apmpackage/apm/data_stream/app_metrics/fields/fields.yml#L30-L35

IIANM the only exception to this is for the `apm.traces` and `apm.rum` data streams, where spans and transactions end up. These both have have `processor.name: transaction`, but they differ in `processor.event` (one is "span", one is "transaction"). Eventually spans and transactions should converge, but for now I think we could set the value in an ingest pipeline.

Maybe we could:
 - set [`event.kind`](https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-kind) to either "span" or "transaction" in the apm-data code
 - update the `traces` ingest pipeline to use this to populate `processor.event`, and then remove `event.kind` since those values are not valid for `event.kind`

WDYT?

Reply:
Doesn't have to be done right now btw, but I think we should do it as part of https://github.com/elastic/apm-data/issues/47

Reply:
> Doesn't have to be done right now btw, but I think we should do it as part of #47

![](https://media2.giphy.com/media/JMV7IKoqzxlrW/giphy.gif?cid=6104955e8q5z8qqvl6w8uou1774zgj70fwsmkmm543k7osx9&ep=v1_gifs_translate&rid=giphy.gif&ct=g)

Reply:
> The "processor" fields are a bit of a relic, and we should aim to remove them in the long term. With that in mind, I wonder if we should change the model a little bit so we can remove them from the apm-data codebase, and set the `processor.*` fields in our ingest pipelines, or by setting a value on `constant_keyword` fields where it makes sense.
> 
> e.g. for metrics, `processor.name` and `processor.event` are both always "metric", so we can update their field definitions to set the value in the mapping: https://github.com/elastic/apm-server/blob/23fb1577909836ebf45e65705df3fd560de5adb1/apmpackage/apm/data_stream/app_metrics/fields/fields.yml#L30-L35
> 
> IIANM the only exception to this is for the `apm.traces` and `apm.rum` data streams, where spans and transactions end up. These both have have `processor.name: transaction`, but they differ in `processor.event` (one is "span", one is "transaction"). Eventually spans and transactions should converge, but for now I think we could set the value in an ingest pipeline.
> 
> Maybe we could:
> 
>     * set [`event.kind`](https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-kind) to either "span" or "transaction" in the apm-data code
> 
>     * update the `traces` ingest pipeline to use this to populate `processor.event`, and then remove `event.kind` since those values are not valid for `event.kind`

I think it's a good plan, we would probably need to coordinate with the kibana team to remove them in the long term to make sure we aren't breaking anything they are relying on